src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -219,3 +220,343 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
     fgMorphBlockStmt(block, stmt DEBUGARG(__FUNCTION__));
     return true;
 }
+
+//------------------------------------------------------------------------
+// optJumpThread: try and bypass the current block by rerouting
+//   flow from predecessors directly to successors.
+//
+// Arguments:
+//   block - block with branch to optimize
+//   domBlock - a dominating block that has an equivalent branch
+//
+// Returns:
+//   True if the branch was optimized.
+//
+// Notes:
+//
+//    A       B          A     B
+//     \     /           |     |
+//      \   /            |     |
+//      block     ==>    |     |
+//      /   \            |     |
+//     /     \           |     |
+//    C       D          C     D
+//
+bool Compiler::optJumpThread(BasicBlock* const block, BasicBlock* const domBlock)
+{
+    assert(block->bbJumpKind == BBJ_COND);
+    assert(domBlock->bbJumpKind == BBJ_COND);
+
+    // If the dominating block is not the immediate dominator
+    // we would need to duplicate a lot of code to thread
+    // the jumps. Pass for now.
+    //
+    if (domBlock != block->bbIDom)
+    {
+        return false;
+    }
+
+    JITDUMP("Both successors of IDom " FMT_BB " reach " FMT_BB " -- attemting jump threading\n", domBlock->bbNum,
+            block->bbNum);
+
+    // Since flow is going to bypass block, make sure there
+    // is nothing in block that can cause a side effect.
+    //
+    // Note we neglect PHI assignments. This reflects a general lack of
```

Comment:
question: What does it cause? Is it a correctness issue or CQ?

Reply:
Good question. I believe it's just CQ, causing us to miss opportunities downstream (say range prop, which chases back through PHIs, will see more viable paths than actually exist). But hard to be 100% confident.

We really should be more explicit about the ways the optimizer relies on stale analysis data and potentially stale IR, and what (if any) update strategies we employ.